include/kokkos-utils/callbacks/TimerListener.hpp

```diff
@@ -0,0 +1,169 @@
+    MatcherType matcher;
+    uint64_t event_id = invalid_event_id;
+};
+
+/**
+ * @brief Listener that starts a timer when a @ref Kokkos::utils::callbacks::PushRegionEvent is received
+ *        and stops it when a corresponding @ref Kokkos::utils::callbacks::PopRegionEvent is received.
+ *
+ * The start event is selected by requiring it to be of type @ref Kokkos::utils::callbacks::PushRegionEvent
+ * and to be a match for @ref matcher. After the start event is received, the listener counts subsequent
+ * @ref Kokkos::utils::callbacks::PushRegionEvent events, corresponding to nested timing regions, and it
+ * decreases this counter when a pop event is received. The timer is stopped when this
+ * counter becomes zero.
+ *
+ * This class is well-suited for timing the execution of a region of code enclosed by
+ * a pair of @c Kokkos::Profiling::pushRegion and @c Kokkos::Profiling::popRegion calls,
+ * or enclosed in a @c Kokkos::Profiling::ScopedRegion.
+ *
+ * @note The use of a counter to track nested regions and stop the timer when the counter
+ *       reaches zero is necessary because @c Kokkos does not provide a way to match
+ *       push and pop region events directly.
+ */
+template <
+    Kokkos::utils::callbacks::Matcher<Kokkos::Impl::type_list<Kokkos::utils::callbacks::PushRegionEvent>> MatcherType,
+    Kokkos::utils::concepts::ExecutionSpace Exec
+>
+class TimerListener<Kokkos::utils::callbacks::PushRegionEvent, Kokkos::utils::callbacks::PopRegionEvent, MatcherType, Exec>
+{
+public:
+    static constexpr uint32_t invalid_nested_level = Kokkos::Experimental::finite_max_v<uint32_t>;
+
+public:
+    template <typename T> requires (std::same_as<std::remove_cvref_t<T>, MatcherType>)
+    TimerListener(Exec exec_, T&& matcher_)
+      : exec(std::move(exec_)), matcher(std::forward<T>(matcher_)) {}
+
+    void operator()(const Kokkos::utils::callbacks::PushRegionEvent& event)
+    {
+        if (matcher(event)) {
+            if (this->connected() || this->closed()) Kokkos::abort("Timer cannot match a region twice.");
+
+            this->timer.reset(this->exec);
+            this->nested_level = 1;
+        } else {
+            if (this->connected()) ++this->nested_level;
+        }
+    }
+
+    void operator()(const Kokkos::utils::callbacks::PopRegionEvent& /* event */)
+    {
+        if (this->connected()) {
+            --this->nested_level;
+
+            if (this->nested_level == 0) this->timer.stop(this->exec);
+        }
+    }
+
+    bool connected() const { return (this->nested_level != invalid_nested_level) && !this->closed(); }
+
+    /// When the timer has matched a pop event corresponding to each nested push event in the matched region,
+    /// if any, as well as to the matched push event, and the stack is thus empty, it is said to be "closed".
+    bool closed() const { return this->timer.is_valid(); }
+
+public:
+    Kokkos::utils::timer::Timer<Exec> timer {};
+
+private:
+    Exec exec {};
+    MatcherType matcher;
```

Comment:
This is non sense to use the timer for an event that has no exec. Use a regular host timer.

Reply:
I changed it, but then I reverted some of the changes. 

The change that remains is that the `Exec` now defaults to `Kokkos::DefaultHostExecutionSpace` for the `RegionTimerListener`.

I reverted because I was thinking of cases where we'd want to time a group of several kernels, not just one, but that all go to a stream. Something like a graph that we'd submit to a stream. And so we may want to do a `PushRegion` before and after `PopRegion` after and record events in the stream.

Some food for thought left for what to do best :)

Reply:
OK. Probably good to add a note in the code.

Reply:
OK. I've added two notes to the doc.